atc/config.go

```diff
@@ -306,6 +306,11 @@ type PlanConfig struct {
 	// corresponds to an Aggregate plan, keyed by the name of each sub-plan
 	Aggregate *PlanSequence `yaml:"aggregate,omitempty" json:"aggregate,omitempty" mapstructure:"aggregate"`
 
+	// a nested chain of steps to run in parallel
+	Parallel *PlanSequence `yaml:"parallel,omitempty" json:"parallel,omitempty" mapstructure:"parallel"`
+	// limit paralelism for a Parallel plan
+	MaxInParallel int `yaml:"max_in_parallel,omitempty" json:"max_in_parallel,omitempty" mapstructure:"max_in_parallel"`
```

Comment:
This name feels awkward but I'm not sure what else to suggest. Maybe it just sounds weird because I'm used to `max_in_flight`. I thought it might make sense to reuse the same name but @clarafu pointed it it might be confusing to have both set - someone could assume `max_in_flight: 3` at the job level is like setting it on all the steps in the plan. Maybe we should reserve 'in flight' terminology for builds, then? :thinking: If so this name is probably fine.

Reply:
I see what you mean, and I agree that it's best to keep `max_in_flight` separate since it could get a bit confusing. Especially when you consider indentation errors in the YAML, which would produce unexpected results if the keyword is recognised in both the job and plan? 

However, right now we have `in_parallel` and `max_in_parallel` - not sure if I like it or not 🤔If I'm thinking outside the box, I would suggest something like this:

- Add a `PlanSequence` in `atc.Config` which we call `steps`.
- Keep `in_parallel` but change it to take a pointer to an `int`.
- Add `in_sequence` and have it take a pointer to a `string`.

We'd use the pointers so that the values would be `nil` if the key was not present in the map, so we can return an error from `validate` if `steps` is specified without either `in_sequence` or `in_parallel`. Then we could represent all three desired functionalities:

```yaml
jobs:
  - name: example
    plan:
    - in_parallel: 3
      steps:
      - task: echo
        config:
         ...
    - in_parallel: # <- We see a 0, and set the limit to the length of the steps. Same as aggregate.
      steps:
      - task: echo
        config:
         ...
    - in_sequence: # <- Equivalent to "do"?
      steps:
      - task: echo
        config:
         ...
```

Might work 🤔

Reply:
I think the last two options there are a little awkward, since knowing how the YAML is parsed, those keys actually point to blank values, and Concourse's parser would have to be careful to notice that they're there at all (since they'd be set to either empty strings or nil - I don't remember).

The first option feels natural but forces you to pick a number in scenarios where you don't really care (unless we support the later two options).

I agree that `in_parallel` and `max_in_parallel` feel a little weird, but it at least allows the common case of "I just want things to run in parallel" to look simple:

```yaml
in_parallel:
- task: echo1
  file: config.yml
- task: echo2
  file: config.yml
- task: echo3
  file: config.yml
```

What if we allow two forms of `in_parallel`?

The first, simplest form which takes no configuration:

```yaml
in_parallel:
- task: echo1
  file: config.yml
- task: echo2
  file: config.yml
- task: echo3
  file: config.yml
```

...and in more complicated scenarios, it's instead an object with `limit` and `steps`:

```
in_parallel:
  limit: 3
  steps:
  - task: echo1
    file: config.yml
  - task: echo2
    file: config.yml
  - task: echo3
    file: config.yml
```

This could also be extended to support `fail_fast: true` (#165).

On the other hand I could see the repetition of `in_parallel` in `max_in_parallel` being a strength of the name, given that they *are* very closely tied together. :man_shrugging: 

Any preference?

Reply:
I think in your example, `in_parallel` would only be able to cover the current `aggregate` and the proposed limited parallelisation, and in that case I would suggest we just stick with `in_parallel` and `max_in_parallel` - mainly because I suspect indentation might trip users up? 

We could still add `fail_fast` to this PR and perhaps throw an error if this key is present in a non-parallel context? However, it seems like the current `atc.Validate()` doesn't really do strict validation for the plan configuration, so perhaps it's fine to just ignore `fail_fast` if it's specified in another context? 🤔